Avoid testing device in cdist when called in a "Math" context

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53973 Restore storage on meta tensors; increase meta coverage
* #54711 Compute Tensor::toString() without reference to backend
* #54710 Make memory overlap a little less precise so it works with null data ptr
* #54709 Make the fuser raise NotImplementedError when unknown device is hit
* **#54708 Avoid testing device in cdist when called in a "Math" context**
* #54707 Make convolution_overrideable default implementation raise NotImplementedError
* #54706 Disambiguate where 'Doesn't run' error message comes from
* #54705 Fix missing format string qualifier
* #54704 Rename linalg.det test so that it generates a valid method name
* #54703 Add empty_generic, which allocates an empty tensor in a device-generic way
* #54597 Fix some more native_functions.yaml mistakes
* #54627 Nasty little hack to preserve NotImplementedError raised in interpreter

cdist advertises itself as Math but actually it error checks that the inputs
are CPU/CUDA in cdist_impl, which is invoked from a composite context in some
situations. I worked around this by ensuring that when cdist_impl was called in
this way, we DON'T do the device checks, but the entire function is a little
janky and I filed an issue about it at #54096

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27338813](https://our.internmc.facebook.com/intern/diff/D27338813)

ezyang committed:
Avoid testing device in cdist when called in a "Math" context

cdist advertises itself as Math but actually it error checks that the inputs
are CPU/CUDA in cdist_impl, which is invoked from a composite context in some
situations. I worked around this by ensuring that when cdist_impl was called in
this way, we DON'T do the device checks, but the entire function is a little
janky and I filed an issue about it at #54096

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented:
@bdhirsh you already reviewed this in #53973  but I split it out after I made more improvements to the top diff

bdhirsh commented on aten/src/ATen/native/Distance.cpp:
```diff
@@ -44,12 +44,11 @@ Tensor _euclidean_dist(const Tensor& x1, const Tensor& x2) {
   TORCH_CHECK(p >= 0, "cdist only supports non-negative p values");
   TORCH_CHECK(device1 == device2, "X1 and X2 must have the same device type. X1: ", device1, " X2: ", device2);
+  // TODO: This is bad; this test should apply universally
   TORCH_CHECK(!x1.is_cuda() || x1.get_device() == x2.get_device(), "device of X1 (", x1.get_device(), ") must match device of X2 (", x2.get_device(), ")");
```
This test looks redundant (and less restrictive) than the one above it, so we can probably just delete it? But you weren't trying to clean up this entire op as part of the PR, so no need to fix that here.

ezyang replied:
`device1 == device2` doesn't subsume because, despite the varnames, it only tests if device types are equal; still need to test indices.

I kind of don't want to invest too much here because I kind of feel structured kernels should probably be responsible for this.

ezyang committed:
Update on "Avoid testing device in cdist when called in a "Math" context"


cdist advertises itself as Math but actually it error checks that the inputs
are CPU/CUDA in cdist_impl, which is invoked from a composite context in some
situations. I worked around this by ensuring that when cdist_impl was called in
this way, we DON'T do the device checks, but the entire function is a little
janky and I filed an issue about it at #54096

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27338813](https://our.internmc.facebook.com/intern/diff/D27338813)

[ghstack-poisoned]
